src/function/gds/output_writer.cpp

```diff
@@ -104,99 +107,139 @@ void PathsOutputWriter::write(processor::FactorizedTable& fTable, nodeID_t dstNo
     curPath.push_back(firstParent);
     auto backtracking = false;
     while (!curPath.empty()) {
-        auto top = curPath[curPath.size() - 1];
-        auto topNodeID = top->getNodeID();
-        if (top->getIter() == 1) {
-            // check path
-            if (checkPathNodeMask(curPath) && checkSemantic(curPath)) {
-                writePath(curPath);
-                fTable.append(vectors);
-            }
+        if (getTop(curPath)->getIter() == 1) {
+            writePath(curPath);
+            fTable.append(vectors);
             backtracking = true;
         }
         if (backtracking) {
-            // This code checks if we should switch from backtracking to forward-tracking, i.e.,
-            // moving forward in the DFS logic to find paths. We switch from backtracking if:
-            // (i) the current top element in the stack has a nextPtr, i.e., the top node has
-            // more parent edges in the BFS graph AND:
-            // (ii.1) if this is the first element in the stack (curPath.size() == 1), i.e., we
-            // are enumerating the parents of the destination, then we should switch to
-            // forward-tracking if the next parent has visited the destination at a length
-            // that's greater than or equal to the lower bound of the recursive join. Otherwise,
-            // we'll enumerate paths that are smaller than the lower; OR
-            // (ii.2) if this is not the first element in the stack, i.e., then we should switch
-            // to forward tracking only if the next parent of the top node in the stack has the
-            // same iter value as the current parent. That's because the levels/iter need to
-            // decrease by 1 each time we add a new node in the stack.
-            if (top->getNextPtr() != nullptr &&
-                ((curPath.size() == 1 && top->getNextPtr()->getIter() >= info.lowerBound) ||
-                    top->getNextPtr()->getIter() == top->getIter())) {
-                curPath[curPath.size() - 1] = top->getNextPtr();
-                backtracking = false;
+            auto next = getTop(curPath)->getNextPtr();
+            while (true) {
+                if (!canNextReplacePathTop(next, curPath)) {
```

Comment:
I believe this code is fine algorithmically but still looks like slow code because of all the functions calls that are being made and the structure of the while loop, continue, and break, seems could be improved. Can you still open an issue to continue optimizing this and assign to me?

Reply:
I have written a fast path, i.e. keep the original one when there is no semantic or predicate.